### ISSUE TITLE
Support for Pre and Post request delegate functionality to support Shape

### ISSUE THREAD

**Issue by seanvancity** [okta/okta-oidc-ios#226]
Both Shape and Okta have SDKs that are opaque to us. 

Okta has a standard SDK where we ask for something at a high level, and the requests are carried out behind the scenes.

Shape has an SDK where it wants to be given the low-level request object prior to it being sent, and be handed the response for further processing after it has been received.

In order to integrate Okta with Shape, we need to grant Shape access to the requests and responses that Okta is generating.

Approach
We ask for two new callbacks to be added to the Okta SDK.

Example:

```
protocol OktaHttpDelegate {
    
    /// Called after request creation, just before send.
    func willSend(request: NSMutableURLRequest)
    
    /// Called after response received, just after receipt.
    func didReceive(response: HTTPURLResponse)
    
}
```

![OktaShapeFlow](https://user-images.githubusercontent.com/20363572/83313186-fccccc80-a1c9-11ea-9ddb-8e5defc76294.jpg)

Will be repeating this issue for the iOS Auth Swift library and similar approach on the equivalent Android SDKs

**Comment by IldarAbdullin-okta** [comment]
Hi @seanvancity ,

There is a way to provide your custom URLSession implementation into the SDK. You have to do the following API call:
```
let youURLSession = URLSession(
            configuration: config,
            delegate: yourDelegateInstance,
            delegateQueue: .main)
OIDURLSessionProvider.setSession(youURLSession)
```
So by providing custom URL session and session delegate you are able to intercept different events from the HTTP stack. For example the following events are available:
```
func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?)
func urlSession(_ session: URLSession, task: URLSessionTask, didSendBodyData bytesSent: Int64, totalBytesSent: Int64, totalBytesExpectedToSend: Int64)
and etc.
```

As an additional example you can also take a look into `OktaOidcConfig.setupURLSession()` function and use it as a reference

Ildar

**Comment by IldarAbdullin-okta** [comment]
Addressed in 3.9.1 release
